### ISSUE TITLE
Custom Trainer results in `missing 1 required positional argument: 'params'`

### ISSUE THREAD

**Issue by rmporsch** [allenai/allennlp#3753]
**System (please complete the following information):**
 - OS: Linux
 - Python version: 3.7
 - AllenNLP version: installed via pip

**Question**
I am trying to implement my own `Trainer` (similar to #3742).

However, I encounter the following problem:
`TypeError: from_params() missing 1 required positional argument: 'params'`

I did some testing and please see below a minimal examples:

```py
@TrainerBase.register("test")
class CustomTrainer(Trainer):
    pass
```

I then run:
```py
from allennlp.commands.train import train_model
from allennlp.common.params import Params
import tempfile
from qsrl.DataSetReader.reader import CustomReader
from qsrl.Trainer.trainer import CustomTrainer
from qsrl.Models.qsrl_model import CustomModel

if __name__ == "__main__":
    params = Params.from_file('configs/experiment.jsonnet')
    serialization_dir = tempfile.mkdtemp()
    model = train_model(params, serialization_dir)
```

the jsonnet looks like this:

```jsonnet
local num_epochs = 20;
local patience = 10;
local batch_size = 5;
local learning_rate = 0.0001;
local min_label = 30;
local bert_model = 'bert-base-uncased';
local mode = "num";

{
    "train_data_path": 'data/english/testing.txt',
    "validation_data_path": 'data/english/testing.txt',
    "dataset_reader": {
        "type": "custom-reader",
        "mode": mode,
        "bert_model_name": bert_model,
    },
    "validation_dataset_reader": {
        "type": "custom-reader",
        "mode": mode,
        "bert_model_name": bert_model,
    },
    "model": {
        "type": "custom-model",
        "bert_model": bert_model,
        "label_smoothing": 0.1
    },
    "iterator": {
        "type": "bucket",
        "sorting_keys": [["tokens", "num_tokens"]],
        'batch_size': batch_size,
        "padding_noise": 0.1,
    },
    "trainer": {
        "type": "test",
        "num_epochs": num_epochs,
        "num_serialized_models_to_keep": 1,
        "optimizer": {
            "type": "adam",
            "lr": learning_rate
        },
        "patience": patience,
        "grad_norm": 10.0,
        "cuda_device": -1,
        "grad_clipping": 5,
    }
}
```

Unfortunately, this results in the above mentioned 
`TypeError: from_params() missing 1 required positional argument: 'params'`

I feel like I am doing something wrong.
Thankful for every help or suggestion.

**Comment by matt-gardner** [comment]
A couple of questions: (1) are you sure you're using version 0.9, and not master?  And (2) can you give a full stack trace?  That particular error is confusing to me without a full stack trace.

**Comment by rmporsch** [comment]
(1) yes, double checked.
(2) sorry should have done that in the first place, here the stack trace:
```
Traceback (most recent call last):
  File "/Users/robertmilanporsch/Library/Preferences/PyCharmCE2019.3/scratches/test_run_allen.py", line 11, in <module>
    model = train_model(params, serialization_dir)
  File "/Users/robertmilanporsch/opt/anaconda3/envs/qsrl_allen/lib/python3.7/site-packages/allennlp/commands/train.py", line 246, in train_model
    trainer = TrainerBase.from_params(params, serialization_dir, recover, cache_directory, cache_prefix)
  File "/Users/robertmilanporsch/opt/anaconda3/envs/qsrl_allen/lib/python3.7/site-packages/allennlp/training/trainer_base.py", line 87, in from_params
    return klass.from_params(params, serialization_dir, recover, cache_directory, cache_prefix)
TypeError: from_params() missing 1 required positional argument: 'params'
```

**Comment by matt-gardner** [comment]
So, this is a bit hard for me to debug, because we've already changed how this works somewhat significantly, and I don't remember at what point different things were added or removed.  Is it possible for you to use master instead of 0.9?  I'd have an easier time helping you with that.

If you really want to stick with 0.9, I *think* the solution is to implement a `from_params` method with the same signature as the `CallbackTrainer`: https://github.com/allenai/allennlp/blob/336de620317db7ea80a8d0a11c7e049b33e5a339/allennlp/training/callback_trainer.py#L296-L300
Except that doesn't have all of the arguments that are getting passed in your stack trace, so I'm not totally sure what's going on...

The short of it is that the way this worked prior to 1.0 was a mess, and we're trying to fix it so it's a lot easier to use.  It still might change a little bit before 1.0 is released, but most of the changes are done.
